Geometry/MTDNumberingBuilder/plugins/DDDCmsMTDConstruction.cc

```diff
@@ -177,3 +204,142 @@ std::unique_ptr<GeometricTimingDet> DDDCmsMTDConstruction::construct(const DDCom
+
+  CmsMTDStringToEnum theCmsMTDStringToEnum;
+
+  CmsMTDConstruction<cms::DDFilteredView> theCmsMTDConstruction;
+
+  std::vector<GeometricTimingDet*> subdet;
+  std::vector<GeometricTimingDet*> layer;
+
+  while (doSubdet) {
+    std::string nodeName(fv.name());
+    GeometricTimingDet::GeometricTimingEnumType fullNode = theCmsMTDStringToEnum.type(nodeName);
+    GeometricTimingDet::GeometricTimingEnumType thisNode =
+        theCmsMTDStringToEnum.type(nodeName.substr(0, CmsMTDStringToEnum::kModStrLen));
+
+#ifdef EDM_ML_DEBUG
+    edm::LogVerbatim("MTDNumbering") << fv.path();
+    edm::LogVerbatim("MTDNumbering") << "Module = " << fv.name() << " fullNode = " << fullNode
+                                     << " thisNode = " << thisNode;
+#endif
+
+    if (fullNode == GeometricTimingDet::BTL || fullNode == GeometricTimingDet::ETL) {
+      // define subdetectors as GeometricTimingDet components
+
+      subdet.emplace_back(theCmsMTDConstruction.buildSubdet(fv));
+    }
+    if (fullNode == GeometricTimingDet::BTLLayer || fullNode == GeometricTimingDet::ETLDisc) {
+      layer.emplace_back(theCmsMTDConstruction.buildLayer(fv));
+#ifdef EDM_ML_DEBUG
+      edm::LogVerbatim("DD4hep_MTDNumbering") << "Number of layers: " << layer.size();
+#endif
+    }
+    if (thisNode == GeometricTimingDet::BTLModule) {
+#ifdef EDM_ML_DEBUG
+      edm::LogVerbatim("DD4hep_MTDNumbering") << "Registered in GeometricTimingDet as type " << thisNode;
+#endif
+      theCmsMTDConstruction.buildBTLModule(fv, layer.back());
+    } else if (thisNode == GeometricTimingDet::ETLModule) {
+#ifdef EDM_ML_DEBUG
+      edm::LogVerbatim("DD4hep_MTDNumbering") << "Registered in GeometricTimingDet as type " << thisNode;
+#endif
+      theCmsMTDConstruction.buildETLModule(fv, layer.back());
+    }
+
+    doSubdet = fv.firstChild();
+  }
+
+  // sort GeometricTimingDet
+
+#ifdef EDM_ML_DEBUG
+  auto comp = mtd->deepComponents();
+  std::stringstream before(std::stringstream::in | std::stringstream::out);
+  for (const auto& it : comp) {
+    before << "ORDER1 " << it->geographicalId().rawId() << " " << it->type() << " " << it->translation().z() << "\n";
+  }
+  edm::LogVerbatim("DD4hep_MTDNumbering") << "GeometricTimingDet order before sorting \n" << before.str();
+#endif
+
+  for (size_t index = 0; index < layer.size(); index++) {
+    GeometricTimingDet::ConstGeometricTimingDetContainer& icomp = layer[index]->components();
+    std::stable_sort(icomp.begin(), icomp.end(), CmsMTDConstruction<cms::DDFilteredView>::mtdOrderZ);
+    std::stable_sort(icomp.begin(), icomp.end(), CmsMTDConstruction<cms::DDFilteredView>::mtdOrderRR);
+    if (index > 0) {
+      std::stable_sort(icomp.begin(), icomp.end(), CmsMTDConstruction<cms::DDFilteredView>::mtdOrderPhi);
+    }
+  }
+
+  // Add layers to subdetectors:
+  // first BTL (one layer only)
+
+  subdet[0]->addComponent(layer[0]);
+
+  // then ETL (number of layers depend on preTDR design or not)
+
+  if (layer.size() == 3) {
```

Comment:
magic numbers also used here
maybe indicates the need for a geometry version tool/function?